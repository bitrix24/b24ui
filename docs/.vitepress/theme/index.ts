/**
 * This is custom theme
 * We change some css
 * @see node_modules/vitepress/dist/client/theme-default/index.js
 */
import { ref, watch, toValue } from 'vue'
import { createRouter, createMemoryHistory } from 'vue-router'
import bitrix24UIPlugin from '@bitrix24/b24ui-nuxt/vue-plugin'
import Layout from './components/ui/Layout.vue'
import ComponentShowExample from './components/ui/ComponentShowExample.vue'
import ComponentProps from './components/ui/ComponentProps.vue'
import ComponentSlots from './components/ui/ComponentSlots.vue'
import ComponentEmits from './components/ui/ComponentEmits.vue'
import Description from './components/ui/Description.vue'
import Placeholder from './components/ui/Placeholder.vue'
import VPBadge from 'vitepress/dist/client/theme-default/components/VPBadge.vue'
import 'vitepress/dist/client/theme-default/styles/fonts.css'
import './tailwind.css'
// region ory theme ////
import './styles/components/custom-block.css'
import './styles/components/vp-code.css'
import './styles/components/vp-code-group.css'
import './styles/components/vp-doc.css'
import './styles/components/vp-sponsor.css'
// endregion ////
import './theme.css'

globalThis.useFetch = async (url: string, options: RequestInit & { params?: any, transform?: (data) => any } = {}) => {
  const data = ref()
  const status = ref('idle')
  async function _fetch() {
    status.value = 'loading'
    try {
      const params = options.params
        ? Object.fromEntries(
            Object.entries(options.params).map(([key, value]) => [key, toValue(value)])
          )
        : {}

      const queryString = new URLSearchParams(params as any).toString()
      const fullUrl = queryString ? `${url}?${queryString}` : url

      data.value = await fetch(fullUrl, options).then(r => r.json()).then(r => options.transform ? options.transform(r) : r)
      status.value = 'success'
    } catch (error) {
      console.error(error)
      status.value = 'error'
    }
  }
  _fetch()

  if (options?.params?.q) {
    watch(
      options.params.q,
      () => {
        _fetch()
      },
      { deep: true }
    )
  }

  return Promise.resolve({
    data,
    status
  })
}

export default {
  Layout,
  enhanceApp: async ({ app }) => {
    // region components ////
    app.component('Badge', VPBadge)
      .component('Description', Description)
      .component('ComponentShowExample', ComponentShowExample)
      .component('ComponentProps', ComponentProps)
      .component('ComponentSlots', ComponentSlots)
      .component('ComponentEmits', ComponentEmits)
      .component('Placeholder', Placeholder)
    // endregion ////

    if (!import.meta.env.SSR) {
      const router = createRouter({
        routes: [
          // { path: '', component: app }
        ],
        history: createMemoryHistory()
        // history: createWebHistory()
      })
      app.use(router)

      app.use(bitrix24UIPlugin)
    }
  }
}
