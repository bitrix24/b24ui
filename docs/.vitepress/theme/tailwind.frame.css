/*
@plugin '@bitrix24/b24style';
@plugin "@bitrix24/b24ui-nuxt";
*/
@import "tailwindcss" source("../../");
@import "@bitrix24/b24ui-nuxt";
/*
*/



