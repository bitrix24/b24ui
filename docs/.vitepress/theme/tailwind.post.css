@import "tailwindcss" source("../../../docs");
@import "@bitrix24/b24ui-nuxt";

/**
 * bxshef: Custom
 * -------------------------------------------------------------------------- */

:root
{
}

.dark
{
  --vp-c-bg: #0f172a;
  --vp-c-bg-alt: #161618;
  --vp-c-bg-elv: #202127;
  --vp-c-bg-soft: #202127;
}

.bg-grid-example {
  background-image: url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 32 32' width='32' height='32' fill='none' stroke='rgb(130 139 149 / 0.25)'%3e%3cpath d='M0 .5H31.5V32'/%3e%3c/svg%3e");
}

.dark\:bg-grid-example:is(.dark *) {
  background-image: url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 32 32' width='32' height='32' fill='none' stroke='rgb(130 139 149 / 0.25)'%3e%3cpath d='M0 .5H31.5V32'/%3e%3c/svg%3e");
}
