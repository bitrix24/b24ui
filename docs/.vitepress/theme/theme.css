/**
 * bxshef: Custom
 * -------------------------------------------------------------------------- */

:root
{
  --sh-scrollbar-thumb: #3c3c43;
  --sh-scrollbar-background: #ebebef;
}

.dark
{
  --vp-c-bg: #0f172a;
  --vp-c-bg-alt: #161618;
  --vp-c-bg-elv: #202127;
  --vp-c-bg-soft: #202127;
  --vp-code-block-bg: #1a2234;

  --sh-scrollbar-thumb: #bdc1c6;
  --sh-scrollbar-background: #3c3c43;
}

.bg-grid-example {
  background-image: url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 32 32' width='32' height='32' fill='none' stroke='rgb(130 139 149 / 0.25)'%3e%3cpath d='M0 .5H31.5V32'/%3e%3c/svg%3e");
}

.dark\:bg-grid-example:is(.dark *) {
  background-image: url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 32 32' width='32' height='32' fill='none' stroke='rgb(130 139 149 / 0.25)'%3e%3cpath d='M0 .5H31.5V32'/%3e%3c/svg%3e");
}

iframe
{
  min-width: 100%;
  border: none;
  scrollbar-width: thin;
  background-color: transparent;
}

/** region scrollbar */
* {
  scrollbar-color: var(--sh-scrollbar-thumb) var(--sh-scrollbar-background);
  scrollbar-width: thin;
}

/* Styling the scrollbar in Chrome and Safari */
*::-webkit-scrollbar {
  width: 6px;
}

/* Set thumb color for Chrome and Safari */
*::-webkit-scrollbar-thumb {
  background-color: var(--sh-scrollbar-thumb);
  border-radius: 4px;
}

/* Set track color for Chrome and Safari */
*::-webkit-scrollbar-track {
  background-color: var(--sh-scrollbar-background);
}
/** endregion */
