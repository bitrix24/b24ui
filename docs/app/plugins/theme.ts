export default defineNuxtPlugin({
  enforce: 'post',
  setup() {
  }
})
