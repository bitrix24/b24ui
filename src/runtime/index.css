@plugin '@bitrix24/b24style';
@import '#build/b24ui.css';
@import './keyframes.css';
@import './air-design-tokens/index.css';

@layer base {
  body,
  .sidebar-layout.--inner {
    /* @see air-design-tokens/???_b24_context*.css */
    --air-theme-background:
      var(--air-theme-bg-image, none)
      var(--air-theme-bg-position, 0 0) / var(--air-theme-bg-size, auto)
      var(--air-theme-bg-repeat, repeat)
      var(--air-theme-bg-attachment, scroll)
      var(--air-theme-bg-color, transparent)
    ;

    --air-theme-background-blurred:
      var(--air-theme-bg-image-blurred, none)
      var(--air-theme-bg-position, 0 0) / var(--air-theme-bg-size, auto)
      var(--air-theme-bg-repeat, repeat)
      var(--air-theme-bg-attachment, scroll)
    ;
  }

  /* @see src/theme/sidebar-layout.ts */
  body {
    background: var(--air-theme-background);
  }

  .sidebar-layout.--inner {
    background: var(--air-theme-background);
  }

  body {
    overflow-y: scroll!important;
    padding-right: 0!important;
    scrollbar-gutter: stable;
  }

  .--app[data-state="loading"] {
    overflow: hidden;
  }
}
