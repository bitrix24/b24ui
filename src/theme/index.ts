export { default as advice } from './advice'
export { default as avatar } from './avatar'
export { default as avatarGroup } from './avatar-group'
export { default as badge } from './badge'
export { default as button } from './button'
export { default as buttonGroup } from './button-group'
export { default as checkbox } from './checkbox'
export { default as chip } from './chip'
export { default as container } from './container'
export { default as countdown } from './countdown'
export { default as formField } from './form-field'
export { default as input } from './input'
export { default as kbd } from './kbd'
export { default as link } from './link'
export { default as progress } from './progress'
export { default as radioGroup } from './radio-group'
export { default as range } from './range'
export { default as separator } from './separator'
export { default as skeleton } from './skeleton'
export { default as switch } from './switch'
export { default as textarea } from './textarea'
export { default as tooltip } from './tooltip'
