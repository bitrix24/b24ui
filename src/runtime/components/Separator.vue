<script lang="ts">
import type { SeparatorProps as _SeparatorProps } from 'reka-ui'
import type { AppConfig } from '@nuxt/schema'
import theme from '#build/b24ui/separator'
import type { AvatarProps, IconComponent } from '../types'
import type { ComponentConfig } from '../types/utils'

type Separator = ComponentConfig<typeof theme, AppConfig, 'separator'>

export interface SeparatorProps extends Pick<_SeparatorProps, 'decorative'> {
  /**
   * The element or component this component should render as.
   * @defaultValue 'div'
   */
  as?: any
  /** Display a label in the middle. */
  label?: string
  /**
   * Display an icon in the middle
   * @IconComponent
   */
  icon?: IconComponent
  /**
   * Display an avatar in the middle
   */
  avatar?: AvatarProps
  /**
   * @defaultValue 'default'
   */
  color?: Separator['variants']['color']
  /**
   * @defaultValue 'thin'
   */
  size?: Separator['variants']['size']
  /**
   * @defaultValue 'solid'
   */
  type?: Separator['variants']['type']
  /**
   * The orientation of the separator.
   * @defaultValue 'horizontal'
   */
  orientation?: _SeparatorProps['orientation']
  class?: any
  b24ui?: Separator['slots']
}

export interface SeparatorSlots {
  default(props?: {}): any
}
</script>

<script setup lang="ts">
import { computed } from 'vue'
import { Separator, useForwardProps } from 'reka-ui'
import { reactivePick } from '@vueuse/core'
import { useAppConfig } from '#imports'
import { tv } from '../utils/tv'
import B24Avatar from './Avatar.vue'

const props = withDefaults(defineProps<SeparatorProps>(), {
  color: 'default',
  orientation: 'horizontal',
  size: 'thin',
  type: 'solid'
})
const slots = defineSlots<SeparatorSlots>()

const appConfig = useAppConfig() as Separator['AppConfig']

const rootProps = useForwardProps(reactivePick(props, 'as', 'decorative', 'orientation'))

const b24ui = computed(() => tv({ extend: tv(theme), ...(appConfig.b24ui?.separator || {}) })({
  color: props.color,
  orientation: props.orientation,
  size: props.size,
  type: props.type
}))
</script>

<template>
  <Separator v-bind="rootProps" :class="b24ui.root({ class: [props.b24ui?.root, props.class] })">
    <div :class="b24ui.border({ class: props.b24ui?.border })" />

    <template v-if="(label !== undefined && label !== null) || icon || avatar || !!slots.default">
      <div :class="b24ui.container({ class: props.b24ui?.container })">
        <slot>
          <span v-if="label !== undefined && label !== null" :class="b24ui.label({ class: props.b24ui?.label })">{{ label }}</span>
          <Component :is="icon" v-else-if="icon" :class="b24ui.icon({ class: props.b24ui?.icon })" />
          <B24Avatar v-else-if="avatar" :size="((props.b24ui?.avatarSize || b24ui.avatarSize()) as AvatarProps['size'])" v-bind="avatar" :class="b24ui.avatar({ class: props.b24ui?.avatar })" />
        </slot>
      </div>

      <div :class="b24ui.border({ class: props.b24ui?.border })" />
    </template>
  </Separator>
</template>
