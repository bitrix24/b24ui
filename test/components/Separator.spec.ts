import { describe, it, expect } from 'vitest'
import { axe } from 'vitest-axe'
import { mountSuspended } from '@nuxt/test-utils/runtime'
import Separator from '../../src/runtime/components/Separator.vue'
import type { SeparatorProps, SeparatorSlots } from '../../src/runtime/components/Separator.vue'
import ComponentRender from '../component-render'
import theme from '#build/b24ui/separator'
import Search2Icon from '@bitrix24/b24icons-vue/main/Search2Icon'

describe('Separator', () => {
  const types = Object.keys(theme.variants.type) as any
  const sizes = Object.keys(theme.variants.size) as any

  it.each([
    // Props
    ['with label', { props: { label: '+1' } }],
    ['with icon', { props: { icon: Search2Icon } }],
    ['with avatar', { props: { avatar: { src: 'https://github.com/bitrix24.png' } } }],
    ['with orientation vertical', { props: { orientation: 'vertical' as const } }],
    ['with decorative', { props: { decorative: true } }],
    ...types.map((type: string) => [`with type ${type}`, { props: { type } }]),
    ...sizes.map((size: string) => [`with size ${size}`, { props: { size } }]),
    ['with color success', { props: { color: 'air-primary-success' } }],
    ['with as', { props: { as: 'span' } }],
    ['with class', { props: { class: 'flex-row-reverse' } }],
    ['with b24ui', { props: { b24ui: { label: 'text-(length:--ui-font-size-lg)' } } }]
  ])('renders %s correctly', async (nameOrHtml: string, options: { props?: SeparatorProps, slots?: Partial<SeparatorSlots> }) => {
    const html = await ComponentRender(nameOrHtml, options, Separator)
    expect(html).toMatchSnapshot()
  })

  it('passes accessibility tests', async () => {
    const wrapper = await mountSuspended(Separator, {
      props: {
        label: '+1',
        icon: Search2Icon
      }
    })

    expect(await axe(wrapper.element)).toHaveNoViolations()
  })
})
