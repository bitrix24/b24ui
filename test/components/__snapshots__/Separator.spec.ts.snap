// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Separator > renders with as correctly 1`] = `"<span data-orientation="horizontal" role="separator" class="flex items-center align-center text-center w-full flex-row"><div class="border-(--ui-color-divider-vibrant-default) w-full border-solid border-t-(length:--ui-border-width-thin)"></div><!--v-if--></span>"`;

exports[`Separator > renders with avatar correctly 1`] = `
"<div data-orientation="horizontal" role="separator" class="flex items-center align-center text-center w-full flex-row">
  <div class="border-(--ui-color-divider-vibrant-default) w-full border-solid border-t-(length:--ui-border-width-thin)"></div>
  <div class="font-[family-name:var(--ui-font-family-primary)] font-(--ui-font-weight-normal) flex text-(--ui-color-design-plain-na-content-secondary) text-(length:--ui-font-size-sm)/(--ui-font-line-height-reset) mx-3 whitespace-nowrap"><span class="air-secondary-accent inline-flex items-center justify-center select-none rounded-full align-middle bg-(--ui-color-base-8) ring ring-(--ui-color-base-7) size-[28px] text-(length:--ui-font-size-xs)/(--ui-font-line-height-reset) shrink-0"><img src="https://github.com/bitrix24.png" width="22" height="22" class="h-full w-full rounded-[inherit] object-cover"></span></div>
  <div class="border-(--ui-color-divider-vibrant-default) w-full border-solid border-t-(length:--ui-border-width-thin)"></div>
</div>"
`;

exports[`Separator > renders with b24ui correctly 1`] = `
"<div data-orientation="horizontal" role="separator" class="flex items-center align-center text-center w-full flex-row">
  <div class="border-(--ui-color-divider-vibrant-default) w-full border-solid border-t-(length:--ui-border-width-thin)"></div>
  <!--v-if-->
</div>"
`;

exports[`Separator > renders with class correctly 1`] = `
"<div data-orientation="horizontal" role="separator" class="flex items-center align-center text-center w-full flex-row-reverse">
  <div class="border-(--ui-color-divider-vibrant-default) w-full border-solid border-t-(length:--ui-border-width-thin)"></div>
  <!--v-if-->
</div>"
`;

exports[`Separator > renders with color success correctly 1`] = `
"<div data-orientation="horizontal" role="separator" class="flex items-center align-center text-center w-full flex-row" color="success">
  <div class="border-(--ui-color-divider-vibrant-default) w-full border-solid border-t-(length:--ui-border-width-thin)"></div>
  <!--v-if-->
</div>"
`;

exports[`Separator > renders with decorative correctly 1`] = `
"<div data-orientation="horizontal" role="none" class="flex items-center align-center text-center w-full flex-row">
  <div class="border-(--ui-color-divider-vibrant-default) w-full border-solid border-t-(length:--ui-border-width-thin)"></div>
  <!--v-if-->
</div>"
`;

exports[`Separator > renders with icon correctly 1`] = `
"<div data-orientation="horizontal" role="separator" class="flex items-center align-center text-center w-full flex-row">
  <div class="border-(--ui-color-divider-vibrant-default) w-full border-solid border-t-(length:--ui-border-width-thin)"></div>
  <div class="font-[family-name:var(--ui-font-family-primary)] font-(--ui-font-weight-normal) flex text-(--ui-color-design-plain-na-content-secondary) text-(length:--ui-font-size-sm)/(--ui-font-line-height-reset) mx-3 whitespace-nowrap"><svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" aria-hidden="true" data-slot="icon" class="shrink-0 size-7">
      <path fill="currentColor" fill-rule="evenodd" d="M14.05 15.886a6.37 6.37 0 1 1 1.836-1.837l3.311 3.311a1 1 0 0 1 0 1.414l-.422.423a1 1 0 0 1-1.414 0zm1.058-5.328a4.55 4.55 0 1 1-9.098 0 4.55 4.55 0 0 1 9.099 0" clip-rule="evenodd"></path>
    </svg></div>
  <div class="border-(--ui-color-divider-vibrant-default) w-full border-solid border-t-(length:--ui-border-width-thin)"></div>
</div>"
`;

exports[`Separator > renders with label correctly 1`] = `
"<div data-orientation="horizontal" role="separator" class="flex items-center align-center text-center w-full flex-row">
  <div class="border-(--ui-color-divider-vibrant-default) w-full border-solid border-t-(length:--ui-border-width-thin)"></div>
  <div class="font-[family-name:var(--ui-font-family-primary)] font-(--ui-font-weight-normal) flex text-(--ui-color-design-plain-na-content-secondary) text-(length:--ui-font-size-sm)/(--ui-font-line-height-reset) mx-3 whitespace-nowrap"><span class="">+1</span></div>
  <div class="border-(--ui-color-divider-vibrant-default) w-full border-solid border-t-(length:--ui-border-width-thin)"></div>
</div>"
`;

exports[`Separator > renders with orientation vertical correctly 1`] = `
"<div data-orientation="vertical" aria-orientation="vertical" role="separator" class="flex items-center align-center text-center h-full flex-col">
  <div class="border-(--ui-color-divider-vibrant-default) h-full border-solid border-s-(length:--ui-border-width-thin)"></div>
  <!--v-if-->
</div>"
`;

exports[`Separator > renders with size thick correctly 1`] = `
"<div data-orientation="horizontal" role="separator" class="flex items-center align-center text-center w-full flex-row">
  <div class="border-(--ui-color-divider-vibrant-default) w-full border-solid border-t-(length:--ui-border-width-thick)"></div>
  <!--v-if-->
</div>"
`;

exports[`Separator > renders with size thin correctly 1`] = `
"<div data-orientation="horizontal" role="separator" class="flex items-center align-center text-center w-full flex-row">
  <div class="border-(--ui-color-divider-vibrant-default) w-full border-solid border-t-(length:--ui-border-width-thin)"></div>
  <!--v-if-->
</div>"
`;

exports[`Separator > renders with type dashed correctly 1`] = `
"<div data-orientation="horizontal" role="separator" class="flex items-center align-center text-center w-full flex-row">
  <div class="border-(--ui-color-divider-vibrant-default) w-full border-dashed border-t-(length:--ui-border-width-thin)"></div>
  <!--v-if-->
</div>"
`;

exports[`Separator > renders with type dotted correctly 1`] = `
"<div data-orientation="horizontal" role="separator" class="flex items-center align-center text-center w-full flex-row">
  <div class="border-(--ui-color-divider-vibrant-default) w-full border-dotted border-t-(length:--ui-border-width-thin)"></div>
  <!--v-if-->
</div>"
`;

exports[`Separator > renders with type double correctly 1`] = `
"<div data-orientation="horizontal" role="separator" class="flex items-center align-center text-center w-full flex-row">
  <div class="border-(--ui-color-divider-vibrant-default) w-full border-double border-t-(length:--ui-border-width-thin)"></div>
  <!--v-if-->
</div>"
`;

exports[`Separator > renders with type solid correctly 1`] = `
"<div data-orientation="horizontal" role="separator" class="flex items-center align-center text-center w-full flex-row">
  <div class="border-(--ui-color-divider-vibrant-default) w-full border-solid border-t-(length:--ui-border-width-thin)"></div>
  <!--v-if-->
</div>"
`;
